File: hade-sdk/android/HadeSDK.kt
package com.hade.sdk

import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

@Serializable
data class HadeDecisionResponse(
    val status: String,
    val decision: DecisionPayload? = null,
    val reasoning: List<String> = emptyList(),
    val confidence: Double = 0.0,
) {
    @Serializable
    data class DecisionPayload(
        val title: String,
        val distance: String,
        val eta: String? = null,
    )
}

class HadeSDK(
    private val baseUrl: String,
    private val client: OkHttpClient = OkHttpClient(),
) {
    suspend fun getDecision(): HadeDecisionResponse = request("initial", null)
    suspend fun regenerate(): HadeDecisionResponse = request("regenerate", null)
    suspend fun refine(tone: String? = null): HadeDecisionResponse = request("refine", tone)

    private fun request(mode: String, refineTone: String?): HadeDecisionResponse {
        val body = buildString {
            append("{" )
            append("\"mode\":\"")
            append(mode)
            append("\"")
            if (refineTone != null) {
                append(",\"signals\":[{\"type\":\"INTENT\",\"content\":\"refine:")
                append(refineTone)
                append("\"}]")
            }
            append("}")
        }.toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url("$baseUrl/hade/decide")
            .post(body)
            .build()

        client.newCall(request).execute().use { response ->
            require(response.isSuccessful) { "HADE request failed: ${response.code}" }
            return kotlinx.serialization.json.Json.decodeFromString(
                HadeDecisionResponse.serializer(),
                response.body!!.string(),
            )
        }
    }
}

File: hade-sdk/swift/HadeSDK.swift
import Foundation

public struct HadeDecisionResponse: Decodable {
    public let status: String
    public let decision: DecisionPayload?
    public let reasoning: [String]
    public let confidence: Double

    public struct DecisionPayload: Decodable {
        public let title: String
        public let distance: String
        public let eta: String?
    }
}

public final class HadeSDK {
    private let baseURL: URL
    private let session: URLSession

    public init(baseURL: URL = URL(string: "https://example.com/api")!, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    public func getDecision() async throws -> HadeDecisionResponse {
        try await request(mode: "initial", refineTone: nil)
    }

    public func regenerate() async throws -> HadeDecisionResponse {
        try await request(mode: "regenerate", refineTone: nil)
    }

    public func refine(tone: String? = nil) async throws -> HadeDecisionResponse {
        try await request(mode: "refine", refineTone: tone)
    }

    private func request(mode: String, refineTone: String?) async throws -> HadeDecisionResponse {
        var request = URLRequest(url: baseURL.appendingPathComponent("hade/decide"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "mode": mode,
            "signals": refineTone == nil ? [] : [["type": "INTENT", "content": "refine:\(refineTone!)"]]
        ])

        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(HadeDecisionResponse.self, from: data)
    }
}
